### PR TITLE
[AutoDiff] Simplify substitution map logic.

### PR DESCRIPTION
**Pull request by dan-zheng**
Simplify `PrimalGenCloner::getOrDiagnoseAssociatedFunctionSubstitutionMap` using
`SubsMap` member and `SubstitutionMap::get` constructor.

### CODE REVIEW

**Review by rxwei — APPROVED** [review]

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -2506,8 +2489,10 @@ bool PrimalGen::performSynthesis(FunctionSynthesisItem item) {
   // Synthesize primal.
   auto substMap = item.original->getForwardingSubstitutionMap();
   if (auto primalGenEnv = item.target->getGenericEnvironment()) {
-    auto primalSubstMap = primalGenEnv->getForwardingSubstitutionMap();
-    substMap = substMap.subst(primalSubstMap);
+    substMap = SubstitutionMap::get(
+        primalGenEnv->getGenericSignature(),
+        QuerySubstitutionMap{primalGenEnv->getForwardingSubstitutionMap()},
```
Comment (slavapestov) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: substMap is equivalent to just getForwardingSubstitutionMap() here

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 2]
```diff
@@ -2314,18 +2311,12 @@ class PrimalGenCloner final
...
-      // If requirement type can be substituted in original substutition map to
-      // form a non-archetype type, use the substituted type.
-      if (auto origFirstType = reqType.subst(origSubstMap))
-        if (!origFirstType->hasArchetype())
-          reqType = origFirstType;
-      // If requirement type has no type parameters and is not an archetype type,
-      // it is valid. Continue.
+      // If requirement type has no type parameters and is not an archetype
+      // type, it is valid. Continue.
       if (!reqType->isTypeParameter() && !reqType->hasArchetype())
```
Comment (slavapestov) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: hasArchetype() will always be false here

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: I mean primalGenEnv->getForwardingSubstitutionMap()

Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Thanks! Logic revamped in https://github.com/apple/swift/pull/22243/commits/0fb115dcc725e543b78d4994ace58eabf8ed69cf. (The old logic was simpler but led to stdlib compilation crash.)

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 2 above)
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Logic revamped in https://github.com/apple/swift/pull/22243/commits/0fb115dcc725e543b78d4994ace58eabf8ed69cf.

**Comment by dan-zheng** [comment]
I believe stdlib compilation failures are fixed, ready for re-review.
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
Merging. I'm happy to address comments later.

**Pull request merged**